Feature/bluetooth connection

senolmurat opened the pull request:
Implemented Bluetooth Live Activities using **IOBluetooth**

### Live Activities
Live activities will be shown when a device connects or disconnects.

![live_activity_connect_1 mov](https://github.com/user-attachments/assets/1783bd1a-28b0-448b-9acf-566189be4f33)


![live_activity_disconnect mov](https://github.com/user-attachments/assets/f312e575-1c36-4214-94ee-f089eaf9ab24)

Connection displays device type on the left and battery percentage on the right with changing colors depending on the percentage (0=red -> 100=green)

- Altough I could not test it, device Icon supports many unique devices like **AirPods, Beats** etc.
- Device Icon is gathered automatically using device name fallbacks to device `MajorClass` and `MinorClass`
- Custom SFSymbol Icons for devices can be set for not recognized devices. 
- Device Icon is tintable with accent color

### Sneak Peek

Sneak Peek is available and can be toggled on/off with 2 different styles, Default and Inline; similar to Media Sneak Peek

![inline_sneak_peek mov](https://github.com/user-attachments/assets/790a0f09-0281-4f3a-bb8c-c033a4944fec)

![standart_sneak_eek_1 mov](https://github.com/user-attachments/assets/20daade9-cd12-4576-a806-b75fc64fb7fa)

### More Examples


![live_activity_connect_2 mov](https://github.com/user-attachments/assets/f3776305-3b11-44a4-9685-ebf2e6c496f1)

![standart_sneak_peek_2 mov](https://github.com/user-attachments/assets/d375f140-61ad-4410-9f13-9cf95072c66e)

![standart_sneak_peek_3 mov](https://github.com/user-attachments/assets/c6d2e5a3-013f-4666-90ae-0b4aa08dce7d)

### Settings
<img width="698" height="597" alt="settings" src="https://github.com/user-attachments/assets/c02c2d02-7a00-445a-92fb-f0c14ab6c5c0" />

### Notes
Any feedback is welcome. 
IOBluetooth is somewhat inconsistent with device information. I have failed to gather battery information consistently on multiple devices and finally I've decided to use `pmset -g accps`  command and captured the output to display battery percentages. 

Device classes are similar stor. Some devices I've tried did not broadcast class info so I opted for name matching and finally custom images so that users can anter a keyword and set a custom image for their unrecognized devices. I've initially used `system_profiler SPBluetoothDataType` command output to gather device `MinorClass` but realized app needs to be un-sandboxed for this command to give output inside an app so I just scraped it as I think it was not worth to un-sandbox this app just for this.

Tried to keep it lightweight its all event driven by Notifications. Only polling is for battery levels when a device connects to wait for `pmset -g accps` output to refresh and that timeouts after 2 seconds

### Issues

- Some devices do not broadcast disconnect signal and disconnect animation does not trigger because of this. I've failed to find a solution for it.
- I've not tested with many devices , I only have a bluetooth keyboard, headphones and xbox gamepad.
- User has to enter the SFSymbol name as text for the custom images. I did not implemented a SFSymbolPicker. I could have used something similar like https://github.com/xnth97/SymbolPicker but I didn't want to bloat it just for this.

senolmurat committed:
bluetooth audio device manager implemented with live activity

senolmurat committed:
small improvements

senolmurat committed:
bluetooth device detection improved

senolmurat committed:
improvements done on battery charge polling

senolmurat committed:
sneak peak fixed, improved bluetooth settings

senolmurat committed:
marqueeText infinite text implemented

senolmurat commented:
Related to #620 and #694
Resolves #620 
Somewhat Resolves #694 , except animation

senolmurat committed:
Merge branch 'dev' into feature/bluetooth_connection

senolmurat committed:
revert package.resolved, display expanding view even if battery information cannot be fetched

Alexander5015 commented:
I think a thinner circle might look better for the battery life, let me know what you think. It may also look better if you rounded the ends of the colored segment.

If you want to use unsandboxed apis, you can update the BoringNotchXPCHelper to make those calls for you, which allows the app to remain sandboxed. If the current solution is working fine, then no need but thought I would point it out. This is being utilized by the HUD already. 

For custom symbols, I think the storage increase is definitely worth it, unless it really adds a lot of extra size to the app. The user experience is the most important thing, there is no need to save a few MB of storage space here imo.

senolmurat committed:
Merge branch 'dev' into feature/bluetooth_connection

senolmurat committed:
thinned out battery ring, added minor class fetching with XPC helper

senolmurat committed:
SymbolPicker implemented

senolmurat committed:
Merge branch 'dev' into feature/bluetooth_connection

senolmurat committed:
bluetooth settings improved, now asks for permission when clicked on settings menu

senolmurat committed:
updated the battery ring

senolmurat committed:
added phone name matching

senolmurat committed:
Update BoringBattery.swift, percentage / 102

senolmurat committed:
Merge branch 'dev' into feature/bluetooth_connection_dev_merge

# Conflicts:
#	BoringNotchXPCHelper/BoringNotchXPCHelper.swift
#	BoringNotchXPCHelper/BoringNotchXPCHelperProtocol.swift
#	boringNotch.xcodeproj/project.pbxproj
#	boringNotch/Localizable.xcstrings
#	boringNotch/XPCHelperClient/BoringNotchXPCHelperProtocol.swift
#	boringNotch/XPCHelperClient/XPCHelperClient.swift
#	boringNotch/boringNotch.entitlements
#	boringNotch/components/Live activities/BoringBattery.swift
#	boringNotch/components/Live activities/MarqueeTextView.swift
#	boringNotch/components/Settings/SettingsView.swift
#	boringNotch/models/Constants.swift

senolmurat committed:
conflicts resolved with the latest dev version

senolmurat committed:
Add Bluetooth connection monitoring, device icon resolution, and battery service functionality

- Implemented BluetoothConnectionMonitor for managing Bluetooth device connections and state changes.
- Created BluetoothDeviceIconResolver to map device snapshots to SF Symbol names for UI representation.
- Added BluetoothPeripheralBatteryService for fetching and polling battery levels of connected Bluetooth devices.
- Introduced BluetoothDeviceSnapshot model to represent the state of Bluetooth devices in the UI.

senolmurat committed:
improvements done

senolmurat committed:
implemented bluetooth icon to the new settings screen
